File: source/mcs_convert/mcs_convert.py

# mcs_convert.py

import sys

def mcs_convert(filename) :
   
   sour_file = "mcsfile_folder/"     + filename
   conv_file = "convert_folder/spi_" + filename
   

   f1 = open(sour_file,"rb")
   f2 = open(conv_file,"wb")
   
   a = range(1000)
   
   counter        = 0
   cmd_start_flag = 0
   
   while 1 :
      for items in range(1000) :
         a[items] = f1.read(1)
         
         if (a[items] == ":") : 
            if (cmd_start_flag == 1) :
               cmd_start_flag = 0
               Byte_Char   = '0x' + a[1] + a[2]
               Record_Type = '0x' + a[7] + a[8]
               
               byte_length = int(Byte_Char,16)
               
               if (Record_Type == '0x00') :
                  for items_in in range(9,9+byte_length*2,2) :
                     b = '0x' + a[items_in] + a[items_in+1]
                     f2.write(chr(int(b,16)))
               
               f1_point = f1.tell()
               f1_start_point = f1_point - 1
               f1.seek(f1_start_point, 0)
               
               break
            elif (cmd_start_flag == 0) :
               cmd_start_flag = 1
            
         
         if (a[items] == "") : 
            if (cmd_start_flag == 1) :
               cmd_start_flag = 0
               Byte_Char   = '0x' + a[1] + a[2]
               Record_Type = '0x' + a[7] + a[8]
               
               byte_length = int(Byte_Char,16)
               
               if (Record_Type == '0x01') :
                  break
               else :
                  print "MCS File Read Error"
                  f1.close()
                  f2.close()
                  sys.exit(-1)
               
         
      
      counter = counter + 1
      if ((counter % 1000) == 0) :
         print counter
      
      if (Record_Type == '0x01') :
         print "All Finished."
         break
         
   
   f1.close()
   f2.close()


if __name__ == '__main__' :
   
   argvs = sys.argv
   argc  = len(argvs)
   
   if (argc != 2) :
      print "Usage: %s <source mcs file name>" % argvs[0]
   
   mcs_convert(filename = argvs[1])
   

File: source/mcs_convert/mcs_convert.cc

// mcs_convert
#include "mcs_convert.h"

using namespace std;

bool mcs_convert(const char *filename)
{

  string mcsfolder="mcsfile_folder/";
  ifstream f1( (mcsfolder+filename).c_str(), ios::in | ios::binary);
  if(!f1){
    cerr << "Cannot open " << (mcsfolder+filename).c_str() << endl;
    return false;
  }
  string convfolder="convert_folder/spi_";
  ofstream f2( (convfolder+filename).c_str(),ios::binary);
  if(!f2){
    cerr << "Cannot open " << (convfolder+filename).c_str() << endl;
    return false;
  }

  const int MAXBYTE=1000;
  char *a = new char[MAXBYTE];

  int counter        = 0;
  bool cmd_start_flag = false;

  while(1)
    {
      char Byte_Char[5]="";
      char Record_Type[5]="";
      int byte_length;

      for(int items=0;items<MAXBYTE;items++)
	{
	  f1.read(&a[items],1);

	  if (a[items] == ':')  {
            if (cmd_start_flag)
	      {
		cmd_start_flag = false;
		strcpy(Byte_Char,"0x");
		strncat(Byte_Char,&a[1],2);
		strcpy(Record_Type,"0x");
		strncat(Record_Type,&a[7],2);

		byte_length = strtol(Byte_Char,0,16);

		if (strcmp(Record_Type, "0x00")==0)
                  for(int items_in=9;items_in<9+byte_length*2;items_in+=2)
		    {
		      char b[5];
		      strcpy(b,"0x");
		      strncat(b,&a[items_in],2);
		      char buf=(char)strtol(b,0,16);
		      f2.write(&buf,1);
		    }

		unsigned int f1_point = f1.tellg();
		unsigned int f1_start_point = f1_point - 1;
		f1.seekg(f1_start_point, ios_base::beg);

		break;
	      }
            else //if (!cmd_start_flag)
	      cmd_start_flag = true;

	  }

	  //if (a[items] == '') 
	  if (f1.eof()) 
            if (cmd_start_flag)
	      {
		cmd_start_flag = false;
		strcpy(Byte_Char,"0x");
		strncat(Byte_Char,&a[1],2);
		strcpy(Record_Type,"0x");
		strncat(Record_Type,&a[7],2);

		byte_length = strtol(Byte_Char,0,16);

		if (strcmp(Record_Type,"0x01")==0)
                  break;
		else{
                  cerr << "MCS File Read Error" << endl;
                  f1.close();
                  f2.close();
                  return false;
		}
	      }               
	}

      counter++;
      if ((counter % 1000) == 0)
	cout << counter << endl;

      if (strcmp(Record_Type, "0x01")==0){
	cout << "All Finished." << endl;
	break;
      }

    }   

  delete[] a;
  f1.close();
  f2.close();

  return true;

}


#ifdef TEST_MCS_CONVERT

int main(int argc,char **argv)
{
  if (argc != 2) {
    cerr << "Usage: "<< argv[0] << " <source mcs file name>" <<endl;
    return 1;
  }
  char *filename=argv[1];

  if(mcs_convert(filename))
    return 0;
  else
    return 1;

}

#endif
